File: Choleski_Method.m



matrix_names = {'ex15.mat', 'apache2.mat', 'cfd1.mat', 'cfd2.mat', 'Flan_1565.mat', 'G3_circuit.mat', 'parabolic_fem.mat', 'shallow_water1.mat', 'StocF-1465.mat'};
%folder = "C:\Users\Simone\Desktop\Università\Magistrale\Metodi del calcolo scientifico\Progetto1\matrici";

if Sys.iswindows()
    folder = "C:\\Users\\Diagon\\Desktop\\UNIMIB\\ANNO 1\\SECONDO SEMESTRE\\Metodi Calcolo\\Matrici-Sparse";
else
    folder = "/home/diagon/Matrici-Sparse";
end

nomi = {};
dimensioni = [];
tempi = [];
errori = [];
memorie = [];

for i = 1 : length(matrix_names)

    data = load(fullfile(folder, matrix_names{i}));

    A = data.Problem.A;
    
    if not(issymmetric(A))
        disp("La matrice non è simmetrica");
    end
    
    disp(['Nome matrice: ' , matrix_names{i}])
    disp([ 'Dimensione matrice : ' ,num2str(size(A,1)), 'x' ,num2str(size(A,2))]);
    
    
    %solo calcolo della memoria
    mem_prima = whos('A');
    L = decomposition(A, 'chol');
    mem_dopo = whos('L').bytes;
    clear L;

    xe = ones(length(A), 1); % Vettore con tutti 1
    b = A * xe;
    
    % Tempo
    tic
    x = A \ b;
    t = toc;
    
    % Errore
    rel_err = norm(x - xe)/norm(xe);
    
    mem_dopo = mem_dopo + whos('x').bytes;
    mem_MB = (mem_dopo - mem_prima.bytes) / 1024^2;

    nomi{end+1} = matrix_names{i};
    dimensioni(end+1) = size(A,1);
    tempi(end+1) = t;
    errori(end+1) = rel_err;
    memorie(end+1) = mem_MB;

    %libero la memoria just to be sure
    clear x;
end

if ispc
    risultati = "risultati_win_matlab.csv";
else
    risultati = "risultati_linux_matlab.csv";
end

T = table(nomi', dimensioni', tempi', errori', memorie', ...
    'VariableNames', {'nome','dimensione','tempo','errore','memoria'});

T = sortrows(T, 'dimensione');

writetable(T, risultati);

File: choleski_method.jl
# Decomposizione di Cholesky in Julia
using LinearAlgebra
using SparseArrays, MAT
using CSV, DataFrames

function cholesky_decomposition()

    if Sys.iswindows()
        folder = "C:\\Users\\Diagon\\Desktop\\UNIMIB\\ANNO 1\\SECONDO SEMESTRE\\Metodi Calcolo\\Matrici-Sparse";
    else
        folder = "/home/diagon/Matrici-Sparse";
    end
    #folder = "matrici\\"
    files = readdir(folder)
    matrici = []
    dimensioni = []
    errori = []
    tempi = []
    memorie = []
    for f in files
        data = matread(joinpath(folder, f))
        if !Sys.iswindows() && f == "Flan_1565.mat" && false
            println("Salto $f: troppo grande per la RAM disponibile (26GB/16GB swap).")
            continue
        else
            push!(matrici, (data["Problem"]["A"], data["Problem"]["name"]))
        end
        
    end


    results = DataFrame(
        nome = String[],
        dimensione = Int[],
        tempo = Float64[],
        errore = Float64[],
        memoria = Float64[]
    )

    #ordino le matrici in base alla dimensione
    matrici = sort(matrici, by = m -> m[1].n)
    dimensioni = [m[1].n for m in matrici]
    println(dimensioni)
    println([(t[2], t[1].n) for t in matrici])

    for (A, A_name) in matrici
        # Controllo se la matrice è simmetrica
        if !(issymmetric(A))
            error("La matrice deve essere simmetrica.")
        end


        println("Matrice: ", A_name)
        # Controllo se la matrice è definita positiva, inutile perchè la sua descrizione è
        # "Test whether a matrix is positive definite (and Hermitian) by trying to perform a Cholesky factorization of A."
        # if !(isposdef(A))
        #     error("La matrice deve essere definita positiva.")
        # end

        # Soluzione esatta xe = [1,1,1,1...]
        xe = ones(size(A, 1))
        mem_prima = Base.summarysize(A) / 1024^2
        println("mem_prima :", mem_prima)
        b = A * xe
        t = @elapsed begin
            
            f = cholesky(A)
            x = f \ b

        end

        s = unsafe_load(f.ptr)

        mem_stimata = (
            (s.xsize * sizeof(Float64) + 
            s.ssize * sizeof(Int64)+ 
            s.nsuper * sizeof(Int64) * 3 +
            length(x) * sizeof(Int64)) / 1024^2
        )

        err = norm(x-xe)/norm(xe)
        #memoria_allocata = mem_stimata - mem_prima
        #push!(errori, err)
        #push!(tempi, t)
        #push!(memorie, memoria_allocata)
        
        push!(results, (
            A_name,
            size(A,1),
            t,
            err,
            mem_stimata - mem_prima
        ))
   end
    filename = Sys.iswindows() ? "risultati_win_julia.csv" : "risultati_linux_julia.csv"
                             
    CSV.write(filename, results)

    return
end

cholesky_decomposition() 
